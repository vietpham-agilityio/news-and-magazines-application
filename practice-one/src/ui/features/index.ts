export * from '@/ui/features/TagGroup';
export * from '@/ui/features/SidebarCard';
export * from '@/ui/features/HashtagGroup';

export { default as BadgeGroup } from '@/ui/features/BadgeGroup';
export { default as CommentGroup } from '@/ui/features/CommentGroup';
export { default as CategoryMenu } from '@/ui/features/CategoryMenu';
export { default as AuthorInformation } from '@/ui/features/AuthorInformation';
export { default as CategoryBurgerMenu } from '@/ui/features/CategoryBurgerMenu';

// sections
export { default as PostsGrid } from '@/ui/sections/PostsGrid';
export { default as PostDetail } from '@/ui/sections/PostDetail';
export { default as RightSidebar } from '@/ui/sections/RightSidebar';
export { default as PostsByCategory } from '@/ui/sections/PostsByCategory';
export { default as PostsByAttribute } from '@/ui/sections/PostsByAttribute';