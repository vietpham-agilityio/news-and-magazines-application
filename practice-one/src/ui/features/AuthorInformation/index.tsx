// services
import { getAuthorById, getPostDataById } from '@/services';

// components
import { Avatar, Button, Information } from '@/ui/components';

// icons
import { Plus } from '@/ui/components/Icons';

interface IProps {
  postId: string;
}

const AuthorInformation = async ({ postId }: IProps) => {
  const postResponse = await getPostDataById(postId);
  const { authorId } = postResponse.data.attributes;

  const { data: authorResponse } = await getAuthorById(authorId);
  const { avatarUrl, name } = authorResponse.attributes;

  return (
    <div className="bg-white-90 flex p-[15px] mb-[25px] rounded-xl">
      <Avatar width={87} height={87} url={avatarUrl} alt="Avatar author" />
      <div className="flex-1 ml-2.5 mt-3.5">
        <Information name={name} information="27 posts" />
        <div className="mt-2">
          <Button
            label="Follow"
            leftIcon={<Plus />}
            bgColor="bg-gradient-to-bl from-sunSet-0 to-sunSet-100"
            textColor="text-white-100"
          />
        </div>
      </div>
    </div>
  );
};

export default AuthorInformation;
