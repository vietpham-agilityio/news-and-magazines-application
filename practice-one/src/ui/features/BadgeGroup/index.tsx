// services
import {
  getCategoryById,
  getCategoryByPostId,
  getCommentByPostId,
  getPostDataById,
} from '@/services';

// componens
import { Tag } from '@/ui/components';

// icons
import { ChatBox, Schedule, File } from '@/ui/components/Icons';

// utils
import { formatDate } from '@/utils';

interface IProps {
  postId: string;
}

export default async function BadgeGroup({ postId }: IProps) {
  const postResponse = await getPostDataById(postId);
  const { data: listComments } = await getCommentByPostId(postId);

  const { data: postCategoriesDataResponse } = await getCategoryByPostId({
    postId,
  });
  const categoryId =
    postCategoriesDataResponse[0]?.attributes.categoryId ?? '1';
  const categoryIdToNumber = parseInt(categoryId, 10);
  const { data: categoryDataResponse } =
    await getCategoryById(categoryIdToNumber);
  const { name: categoryType } = categoryDataResponse.attributes;

  const attributes = postResponse?.data?.attributes;
  const publicDate = formatDate(attributes?.publicationDate) ?? '';
  const commentCount = listComments?.length ?? '0';

  return (
    <div className="flex w-full justify-evenly mx-auto mb-12 xl:w-2/4">
      <Tag icon={<Schedule />} content={publicDate} />
      <Tag
        icon={<ChatBox />}
        content={`Comments: ${commentCount}`}
        additionalClass="hidden sm:flex"
      />
      <Tag
        icon={<File />}
        content={`Categories: ${categoryType}`}
        additionalClass="hidden sm:flex"
      />
      <Tag
        icon={<ChatBox />}
        content={`${commentCount}`}
        additionalClass="sm:hidden"
      />
      <Tag
        icon={<File />}
        content={`${categoryType}`}
        additionalClass="sm:hidden"
      />
    </div>
  );
}
