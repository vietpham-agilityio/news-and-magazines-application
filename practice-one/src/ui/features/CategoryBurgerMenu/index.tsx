// services
import { getCategoryData } from '@/services';

// types
import { ICategoryItem, IMenu } from '@/types';

// componens
import { BurgerMenu } from '@/ui/components';

const CategoryBurgerMenu = async () => {
  const { data: categoryDataResponse = [] } = await getCategoryData();

  const listCategories: IMenu[] = [];

  categoryDataResponse?.map((category: ICategoryItem) => {
    const { id } = category;
    const { name } = category.attributes;

    listCategories.push({
      id: id.toString(),
      name: name,
    });
  });

  return <BurgerMenu listCategories={listCategories} />;
};

export default CategoryBurgerMenu;
