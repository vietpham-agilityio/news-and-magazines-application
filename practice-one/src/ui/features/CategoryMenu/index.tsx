// services
import { getCategoryData } from '@/services';
import { ICategoryItem, IMenu } from '@/types';

// componens
import { Menu } from '@/ui/components';

export default async function CategoryMenu() {
  const { data: categoryDataResponse = [] } = await getCategoryData();

  const listCategories: IMenu[] = [];

  categoryDataResponse.map((category: ICategoryItem) => {
    const { id } = category;
    const { name } = category.attributes;

    listCategories.push({
      id: id.toString(),
      name: name,
    });
  });

  return <Menu name="Categories" listMenu={listCategories} />;
}
