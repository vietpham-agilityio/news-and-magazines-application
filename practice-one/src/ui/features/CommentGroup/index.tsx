// components
import { Comment, Typography } from '@/ui/components';

// services
import { getCommentByPostId } from '@/services';

// types
import { FontWeight } from '@/types';

// utils
import { formatDate } from '@/utils';

interface IProps {
  postId: string;
}

const CommentGroup = async ({ postId }: IProps) => {
  const { data: listComments } = await getCommentByPostId(postId);

  if (listComments.length === 0) {
    return (
      <div className="flex justify-center py-10">
        <Typography
          tag="h3"
          weight={FontWeight.Light}
          additionalClasses="text-sm text-black-75 leading-5"
        >
          This article has no comments
        </Typography>
      </div>
    );
  }

  return (
    <div className="flex flex-col gap-9 mb-10 sm:mb-12 lg:mb-15">
      {listComments.map(comment => {
        const { name, content, createdAt } = comment.attributes;
        return (
          <Comment
            key={comment.id}
            userName={name}
            content={content}
            publicDate={formatDate(createdAt)}
          />
        );
      })}
    </div>
  );
};

export default CommentGroup;
