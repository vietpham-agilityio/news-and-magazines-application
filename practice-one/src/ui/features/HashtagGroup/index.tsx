// types
import { IHashtag } from "@/types";
import { LazyImage } from "@/ui/components";

interface IProps {
  listHashtags: IHashtag[];
}

export const HashtagGroup = ({ listHashtags }: IProps) =>
  listHashtags.map((hashtag) => {
    const { id, name, imageUrl } = hashtag;
    return (
      <div key={id} className="mx-3 py-2.5 rounded-xl cursor-pointer">
        <LazyImage
         src={imageUrl}
         alt={name}
         width={170}
         height={48}
        />
      </div>
    );
  });
