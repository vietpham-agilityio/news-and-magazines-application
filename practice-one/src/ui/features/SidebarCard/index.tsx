// components
import { Title, PaperPost } from '@/ui/components';

// types
import { PostResponse, PostVariant } from '@/types';

// services
import { getPostDataByAttribute } from '@/services';

export const SidebarCard = async () => {
  const { data: postDataResponse } = await getPostDataByAttribute({
    attribute: PostVariant.TOP,
    limit: 6,
  });
  return (
    <div className="px-[15px] py-5 bg-white-90 rounded-xl">
      <Title title="Top Posts" />
      <div className="mt-5">
        {postDataResponse.map((post: PostResponse, index) => {
          const { title, content, imageUrl } = post.attributes;
          return (
            <div
              key={post.id}
              className={`${index !== postDataResponse.length - 1 && 'mb-5'}`}
            >
              <PaperPost
                id={post.id}
                title={title}
                content={content}
                imageSrc={imageUrl}
                alt={title}
              />
            </div>
          );
        })}
      </div>
    </div>
  );
};
