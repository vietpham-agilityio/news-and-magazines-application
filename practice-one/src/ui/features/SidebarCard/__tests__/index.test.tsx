import { render, waitFor } from '@testing-library/react';

// service
import { getPostDataByAttribute } from '@/services';

// component
import { SidebarCard } from '@/ui/features';
import { mockPostsByAttributeData } from '@/mocks';

jest.mock('../../../../services', () => ({
  ...jest.requireActual('../../../../services'),
  getPostDataByAttribute: jest.fn(),
}));

beforeEach(() => {
  (getPostDataByAttribute as jest.MockedFunction<typeof Object>).mockResolvedValue(
    mockPostsByAttributeData
  );
});

afterEach(() => {
  jest.clearAllMocks();
});

describe('SidebarCard component', () => {
  test('Should render match with snapshot.', async () => {
    const { container } = render(await SidebarCard());

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
