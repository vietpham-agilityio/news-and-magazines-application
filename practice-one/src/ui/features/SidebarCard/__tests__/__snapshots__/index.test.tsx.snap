// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarCard component Should render match with snapshot. 1`] = `
<div>
  <div
    class="px-[15px] py-5 bg-white-90 rounded-xl"
  >
    <div
      class="flex items-center"
    >
      <div
        class="w-1 h-[11px] bg-primary-100 rounded-2xl mr-1.5"
      />
      <h2
        aria-level="2"
        class="text-md font-medium text-dark-100 leading-6"
        role="heading"
      >
        Top Posts
      </h2>
    </div>
    <div
      class="flex flex-col justify-between mt-5 h-155.5"
    >
      <div>
        <a
          href="/articles/12"
        >
          <div
            class="flex cursor-pointer undefined"
          >
            <div
              style="width: 87px; height: 87px; position: relative;"
            >
              <img
                alt="Balancing Innovation and Responsibility"
                class="rounded-xl"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                sizes="87px"
                src="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=3840&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=16&q=75 16w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=32&q=75 32w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=48&q=75 48w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=64&q=75 64w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=96&q=75 96w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=128&q=75 128w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1608904063217-4e24362cda27%3Fw%3D800%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxzZWFyY2h8MTB8fEJhbGFuY2luZyUyMElubm92YXRpb24lMjBhbmQlMjBSZXNwb25zaWJpbGl0eXxlbnwwfHwwfHx8MA%253D%253D&w=3840&q=75 3840w"
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
              />
            </div>
            <div
              class="flex flex-col gap-1.5 ml-2.5 flex-1  my-3"
            >
              <h2
                aria-level="2"
                class="text-xs font-medium w-[223px] text-dark-100 text-ellipsis  undefined"
                role="heading"
              >
                Balancing Innovation and Responsibility
              </h2>
              <h3
                aria-level="3"
                class="text-md font-light 
            h-8 w-[223px] text-dark-75 text-wrap truncate text-ellipsis overflow-hidden
            text-p
            undefined"
                role="heading"
              >
                Artificial Intelligence (AI) technologies have the potential to revolutionize industries, streamline processes, and improve lives. However, the rapid advancement and deployment of AI also raise significant ethical considerations regarding fairness, accountability, transparency, and privacy. This article examines the ethical implications of AI, the challenges they present, and strategies for promoting responsible AI development and deployment
              </h3>
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;
